### PR TITLE
cmake: fix rpath on shared libraries and binaries targets

### PR DESCRIPTION
**Pull request by rjfd**
While building the Ceph RPMs in opensuse build system, we got several errors as mention in the end of the discussion in #12840 

The problem was due to the attempt of CMake to rewrite the RPATH of the binaries, or shared libraries, that didn't have any RPATH linked during the building phase.

Signed-off-by: Ricardo Dias <rdias@suse.com>

### CODE REVIEW

**Comment by smithfarm** [comment]
@rjfd Can you push this wip branch to ceph/ceph-ci.git so Shaman will build it?

**Comment by rjfd** [comment]
@smithfarm done. The branch in ceph-ci repo with this fix is wip-rdias-testing

**Comment by rjfd** [comment]
http://pulpito.ceph.com/rdias-2017-01-14_12:18:42-rados-wip-rdias-testing---basic-smithi/

**Comment by tchaikov** [comment]
@rjfd @smithfarm i'd suggest pass `-DCMAKE_BUILD_WITH_INSTALL_RPATH=ON` to cmake in the ceph.spec recipe a try, as i found that this change is not only changing the `INSTALL_RPATH` of almost all shared libraries and some executables. if we are not performing tests in the built path, we can just build the libraries and executables with the `INSTALL_RPATH`, this approach

- avoids the relink at the install-time.
- is a more consistent way to fix this issue, and
- has lower maintenance overhead

what do you think?

**Comment by rjfd** [comment]
@tchaikov I've tried to build with `CMAKE_BUILD_WITH_INSTALL_RPATH=ON` and the build failed due to undefined references. It looks like that the build of some libraries, such as libcephfs requires that libceph-common to be found, and since RPATH is set to the INSTALL directory, it cannon find it during compilation/linking.

That is why I used the fix submitted in this PR.

File: src/tools/CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +22,8 @@ install(TARGETS ceph_scratchtoolpp DESTINATION bin)
 
 add_executable(ceph_radosacl radosacl.cc)
 target_link_libraries(ceph_radosacl librados global)
+set_target_properties(ceph_radosacl PROPERTIES
```
Comment (tchaikov) [inline src/tools/CMakeLists.txt]: @rjfd why you don't set the `INSTALL_RPATH` for all executables? and what makes `ceph_radosacl`, `ceph_scratchtool` and `ceph_test_libcephfs` different from others?

Comment (rjfd) [inline src/tools/CMakeLists.txt]: you're right `ceph_radosacl`, `ceph_scratchtool`, and `ceph_test_libcephfs` don't need to set the INSTALL_RPATH property. I wrongly identified them.

> why you don't set the `INSTALL_RPATH` for all executables?
I don't understand what you mean, currently `INSTALL_RPATH` is already set for all binaries, what I'm currently doing is setting an empty 'INSTALL_RPATH` for all binaries that don't depend on any internal (also being built) target

Comment (tchaikov) [inline src/tools/CMakeLists.txt]: @rjfd got it. i just wanted to understand the rationale of your change.

**Comment by rjfd** [comment]
@tchaikov pushed the new changes addressing your comment

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -914,6 +918,8 @@ if(WITH_RBD)
   if(WITH_KRBD)
     add_library(krbd STATIC krbd.cc
       $<TARGET_OBJECTS:parse_secret_objs>)
+    set_target_properties(krbd PROPERTIES
```
Comment (tchaikov) [inline src/CMakeLists.txt]: krbd is an convenient library, and we don't even install it. why would we need to set rpath for it?

Comment (rjfd) [inline src/CMakeLists.txt]: You're right. Will fix this.

**Comment by rjfd** [comment]
@tchaikov pushed new change.

**Comment by tchaikov** [comment]
@rjfd, could you update the commit message explaining your change in more details :

for example:

> The problem was due to the attempt of CMake to rewrite the RPATH of the binaries, or shared libraries, that didn't have any RPATH linked during the building phase.

> currently `INSTALL_RPATH` is already set for all binaries, what I'm currently doing is setting an empty 'INSTALL_RPATH` for all binaries that don't depend on any internal (also being built) target

**Comment by rjfd** [comment]
@tchaikov done!

**Pull request merged**